content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md

```diff
@@ -0,0 +1,152 @@
+---
+authors: [tom]
+title: "Null-safety annotations using Spring Boot"
+categories: ["Spring Boot"]
+date: 2022-01-16T00:00:00
+excerpt: "A guide on how to write null-safe code using Spring annotations"
+url: spring-boot-null-safety-annotations
+---
+
+## Introduction
+
+[`NullPointerException`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/NullPointerException.html) (often shortened as NPE) is a nightmare for every Java programmer.
+
+We can find plenty of articles on the internet explaining how to write null-safe code. Null-safety ensures that we have added proper checks in the code to **guarantee the object reference cannot be null or possible safety measures are taken beforehand when any such situation arises**.
+
+Since `NullPointerException` is a runtime exception, it would be hard to figure out such cases during code compilation. Java's type system does not have a way to quickly eliminate the dangerous null object references. 
+
+But, what if I tell you there are helpful approaches to write null-safe code. In particular, if you are using the Spring Framework. Wouldn't that be great news?
+
+Luckily, Spring Framework offers some annotations to solve this exact problem. 
+In this article, we will learn how to use these annotations to write null-safe code using [Spring Boot](https://reflectoring.io/categories/spring-boot/).
+
+## Null-Safety Annotations in Spring
+
+Under the `org.springframework.lang` Spring core package, there are 4 such annotations, `@NonNull`, `@NonNullFields`, `@Nullable`, and `@NonNullApi`. These are [meta-annotations](https://dzone.com/articles/what-are-meta-annotations-in-java) that use the famous (even though dormant at present) [JSR-305](https://jcp.org/en/jsr/detail?id=305) annotations underneath. 
```

Comment:
Why are these annotations meta annotations? We're using them directly on our code later, which means they're not just meta annotations. I removed the sentence, but feel free to re-add it, but with an explanation.